lib/compiler/shared/ast.ts

```diff
@@ -0,0 +1,85 @@
+export type ParentNode = RootNode | ElementNode | ComponentNode | CustomElementNode | ExpressionNode;
+export type Node = RootNode | ElementNode | ComponentNode | CustomElementNode | ExpressionNode | TextNode | FrontmatterNode | DoctypeNode | CommentNode;
+
+export interface Position {
+  start: Point;
+  end?: Point;
+}
+export interface Point {
+  /** 1-based line number */
+  line: number;
+  /** 1-based column number, per-line */
+  column: number;
+  /** 0-based byte offset */
+  offset: number;
+}
+export interface BaseNode {
+  type: string;
+  position?: Position;
+}
+
+export interface ParentLikeNode extends BaseNode {
+  type: 'element' | 'component' | 'custom-element' | 'expression' | 'root';
+  children: Node[];
+}
+
+export interface LiteralNode extends BaseNode {
+  type: 'text' | 'doctype' | 'comment' | 'frontmatter';
+  value: string;
+}
+
+export interface RootNode extends ParentLikeNode {
+  type: 'root';
+}
+
+export interface AttributeNode extends BaseNode {
+  type: 'attribute';
+  kind: 'quoted' | 'empty' | 'expression' | 'spread' | 'shorthand' | 'template-literal';
+  name: string;
+  value: string;
+}
+
+export interface DirectiveNode extends Omit<AttributeNode, 'type'> {
+  type: 'directive';
+}
```

Comment:
One choice worth calling out—I've split `attributes` and `directives` into two separate nodes. This should make it easy to find and validate official directive without needing to loop through all the existing attributes. The compiler filters them out for you.